Summarizer cancellable

vladsud opened the pull request:
Issue #6896: **Summarizer keep going when summarizer container is closed**

**Key changes:**

1. Introduce ICancellable interface to represent cancellation token. It's similar to AbortSignal, but easier to work with as promises compose easier.
2. Consolidate all cancellation logic in coordinator object (RunWhileConnectedCoordinator) that implements ICancellable. Before this change various layers checked for being connected or other conditions, and in many cases - we did not check it at all as proper object was not available at given layer.
   - While it currently does not change a fact that summarizing client needs to be connected to ordering service to proceed with summaries, this might change in the future (with single-commit summaries), so having one source of truth for decision making on continue or not makes design more flexible for future changes.
   - Similar consolidation is done RE telemetry - coordinator is now responsible for start/end telemetry events.
3. Propagate ICancellable through all summarization layers and start checking cancellation (including using Promise.race wherever we wait for promises) to ensure client moves to exit immediately on cancellation. This creates proper infrastructure to cancel summary (using right trigger) and avoid in the future condition of two clients submitting summaries and stepping on each other toes. This at the moment does not reduce number of collisions due to presence of "last Summary" that by design creates such overlap, but I'll improve here in further PRs. This change is especially important in closing gaps like
   - RunningSummarizer.trySummarize() doing 3 summary attempts with delays and not checking for cancellation
   - Waiting for summary op, summary ack/nack
   - Later results in summary timeouts from a client that was closed 10 minutes ago, creating a lot of noise in scalability tests (and I assume in production as well).
4. Stop checking for onBehalfOfClient as it does not really work.
   - Current design is that onBehalfOfClient is updated whenever parent container clientId changes. This makes this check not very useful, as it creates a condition for lastSummary to be cancelled right away, and it indeed happens very often as we observe with documents that have 10K of unsummarized ops and have really hard time to do such summary (in presence of main client continuously getting nacked and disconnected).
   - Ideal solution in this space would for current connected summarizer client to be elected as summarizer parent (i.e. self == parent) even if its actual parent has been disconnected/disposed.
   - This change results in more chances that two summarizer clients are running at the same time and stepping on each other toes, but this situation resolves pretty quickly with old summarizer moving to exit after an attempt. I'll introduce some improvement in next PR in this space to reduce such cases to let last summary run more freely alone.

There are other less interesting changes in here like
- better detection of disposed / disconnected summarizer client before it even started the work.

**Next is series**: Adjust logic around when summarizer client starts to reduce substantially overlap with last summary and reduce to a minimum cases where we have 2 summarizers at the same time.

vladsud committed:
Cancellable summarizer

vladsud committed:
submitSummaryCallback

vladsud committed:
Some cleanup

vladsud commented on packages/runtime/container-runtime/src/runningSummarizer.ts:
```diff
@@ -177,16 +176,7 @@ export class RunningSummarizer implements IDisposable {
 
     public handleSystemOp(op: ISequencedDocumentMessage) {
         switch (op.type) {
-            case MessageType.ClientLeave: {
```
Note RE this change: I tried to understand it better by replacing it with assert and running through various tests, including stress tests - I was never been able to hit it. I think it's due to a fact that before latest change with noops, client may have been removed from room (receive leave op for itself) on 5 min idle interval, and in such state connection would not be torn down but would downgrade to "read" without runtime layers understanding it.
This no longer can happen with more frequent noops.
If we get into same situation again, I'd argue the better fix is to ensure that all layers are aware that client lost "write" connection by cycling through synthetized disconnect/reconnect transitions. That way summarizer would know that it's cancelled. Just directly close connection if DeltaManager_reconnectMode !== ReconnectMode.Enabled.

agarwal-navin replied:
How will we know if we get into this situation? Should we keep the assert that you tried with or add telemetry?

vladsud replied:
I do not think this code should care about it. I'll open an issue to consider these transitions at loader level - I think it's fine (for starters) to log an error that client should never receive its own leave message (I'll try if it's the case) and connection should be closed instead first. 
I'll add it to already opened https://github.com/microsoft/FluidFramework/issues/7180 that tracks it from different angle (as observed by ordering service).

anthony-murphy commented on packages/runtime/container-runtime/src/summarizerTypes.ts:
```diff
@@ -34,6 +33,12 @@ export interface IProvideSummarizer {
+/* Similar to AbortSignal, but using promise instead of events */
+export interface ICancellable {
+    readonly cancelled: boolean;
+    readonly waitCancelled: Promise<void>;
```
can you add comments? i don't understand this promise? will it always resolve? should it be waitFinished or something?

vladsud replied:
I'll add comments. It is resolved when this object is cancelled.

anthony-murphy commented on packages/runtime/container-runtime/src/summarizerTypes.ts:
```diff
@@ -34,6 +33,12 @@ export interface IProvideSummarizer {
     readonly ISummarizer: ISummarizer;
 }
 
+/* Similar to AbortSignal, but using promise instead of events */
```
why are promises better than events? i think i'd prefer to just have an interface that maps to abort signal:
https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal

it's pretty trivial to add a wrapper over the event if you want a promise.

vladsud replied:
Promises compose easier. I.e. 
Promise.race([ something I wait for, cancellable.waitCacenlled]);

That same thing becomes a lot of lines of code if I were using signals.
At the end of the day, I'd write converter from abort signal to promise and would use it :)

anthony-murphy commented on packages/runtime/container-runtime/src/test/runningSummarizer.spec.ts:
```diff
@@ -128,58 +127,59 @@ describe("Runtime", () => {
+                    // ICancellable
+                    {
+                        cancelled: false,
+                        waitCancelled: new Promise(() => {}),
```
isn't this a memory leak? this promise will never resolve or reject

anthony-murphy replied:
anyone who waits on it will wait forever

vladsud replied:
As any other object, it would get collected when there are no references to it. It will only create forever leak if code using it never finishes (for other reasons), which would be a bug to investigate / fix.

vladsud replied:
That's by design - this path is never cancelled. So if someone waits for cancelation, then it will never happen.
It's no different than let's say we added cancelation signal to all our fetches, and test code would pass signal that never gets activated - it's the right mechanism, just happened that this mechanism is never engaged in this (test) flow.

Maybe other way to say it - this flow does not test cancellation, thus it uses promise that never resolves or rejects.

vladsud committed:
Small last minute mistake in binding; comments; policy check

vladsud commented:
Note: I'm thinking to rename ICancellable to ICancelationToken - term more clearly defining purpose (IMHO)

vladsud committed:
Merge branch 'main' of https://github.com/microsoft/FluidFramework into Summarizer_cancellable

# Conflicts:
#	api-report/container-runtime.api.md
#	packages/runtime/container-runtime/src/runWhileConnectedCoordinator.ts
#	packages/runtime/container-runtime/src/summarizer.ts
#	packages/runtime/container-runtime/src/summarizerTypes.ts
#	packages/runtime/container-runtime/src/summaryManager.ts

vladsud committed:
Build breaks & small adjustments

vladsud committed:
ICancellable -> ICancellationToken, and appropriate renaming of variables

vladsud committed:
Merge branch 'main' of https://github.com/microsoft/FluidFramework into Summarizer_cancellable

# Conflicts:
#	packages/runtime/container-runtime/src/runningSummarizer.ts
#	packages/runtime/container-runtime/src/summaryGenerator.ts

vladsud committed:
common-utils

vladsud committed:
Merge branch 'main' of https://github.com/microsoft/FluidFramework into Summarizer_cancellable

vladsud committed:
1. Cancelation token to return a reason for cancelation
2. Address earlier regression where all summaires (including last summary) were cancelled on main client being disconnected. Address by separating summarizer cancellation from cancellation token transitions, similar to how it was before.
Opened https://github.com/microsoft/FluidFramework/issues/7292 to track addition of UTs for Summarizer class.

vladsud committed:
Naming: neverCancelledToken -> neverCancelledSummaryToken

agarwal-navin commented on packages/runtime/container-runtime/src/runWhileConnectedCoordinator.ts:
```diff
@@ -59,38 +89,21 @@ export class RunWhileConnectedCoordinator {
-                return { started: false, message: "DisconnectedBeforeRun" };
-            }
+    public async waitStart() {
+        if (!this.runtime.connected && !this.everConnected) {
```
We immediately return if we are connected or if we were connected sometime before but aren't connected anymore. How does a caller distinguish these two cases? Also, the promise returned can be resolved if cancelled.
Before this change, the return value could have been used. But now, does the caller have to check for connected after this call before proceeding further?

vladsud replied:
The caller should never check for connected state - that is abstracted away by this call (cancellation token abstraction).
And yes, users of the system (caller and all the other layers that get cancellation token) need to check for cancellation frequently in various points to ensure that process moves to exit quickly on cancellation. waitStart() is not special here in any form - it's only special in a sense that this object needs waitStart() to be called to be functional.

Here is current usage :

```typescript
        const runCoordinator = new RunWhileConnectedCoordinator(this.runtime);
        await runCoordinator.waitStart();

        // Wait for either external signal to cancel, or loss of connectivity.
        const stopP = Promise.race([runCoordinator.waitCancelled, this.stopDeferred.promise]);
        void stopP.then((reason) => {
            this.logger.sendTelemetryEvent({
                eventName: "StoppingSummarizer",
                onBehalfOf,
                reason,
            });
        });

        if (runCoordinator.cancelled) {
            return;
        }
```

vladsud replied:
Basically there is only one way of doing it, not two as before - waitStart() is no longer unique place that caller needs to worry about (other than need to call it).

agarwal-navin replied:
Okay. So, waitStart() is basically something that you need to call the first time after you create the coordinator. When it completes, the client is either connected or it was cancelled. The caller doesn't care about connected, it just needs to check that it is not cancelled before it proceeds.

vladsud replied:
Ok, removing / hiding it allocher, by exposing static async create() method. It's much better now :)

agarwal-navin replied:
Okay cool. I was merely pointing out my understanding though :-)
Also nit: waitStart() should eventually become a private method.

agarwal-navin commented on packages/runtime/container-runtime/src/runWhileConnectedCoordinator.ts:
```diff
@@ -59,38 +89,21 @@ export class RunWhileConnectedCoordinator {
      * The promise will also resolve if stopped either externally or by disconnect.
      * The return value indicates whether the start is successful or not.
```
Remove this part of the comment as this function doesn't return anything anymore.

agarwal-navin commented on packages/runtime/container-runtime/src/runWhileConnectedCoordinator.ts:
```diff
@@ -3,54 +3,84 @@
             this.everConnected = true;
-        } else {
+        } else if (this.runtime.disposed) {
+            this.stop("summarizeClientDisconnected");
```
nit: summarizeClientDisconnected -> summarizerClientDisconnected.

vladsud replied:
I noticed that too :) I'll rename in separate PR not to pile more stuff into already big PR.

vladsud committed:
PR feedback - comment

vladsud committed:
Removing waitStart() from public API
